drivers/sensor/bma4xx/bma4xx_emul.c

```diff
@@ -0,0 +1,428 @@
+ */
+#include "bma4xx_regs.h"
+#include "bma4xx_emul.h"
+#include "zephyr/toolchain.h"
+
+#include <errno.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/emul.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/i2c_emul.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/math_extras.h>
+
+#define DT_DRV_COMPAT bosch_bma4xx
+
+LOG_MODULE_REGISTER(bosch_bma4xx, LOG_LEVEL_INF);
+
+/* The total number of used registers specified in bma4xx datasheet. */
+#define NUM_REGS 0x7E
+
+struct bma4xx_emul_data {
+	/* Holds register data. */
+	uint8_t regs[NUM_REGS];
+	/** True if the sensor is currently enabled. */
+	bool accel_enabled;
+	/** Current sensor range, ±2/4/8/16g; value is positive gs. */
+	uint8_t accel_range;
+	/** Raw register value of ACC_CONF:acc_odr. */
+	uint8_t odr_raw;
+	/** Current sensor reading on XYZ axes, in milli-g. */
+	int acceleration[3];
+	/** Axis offset register values, XYZ. */
+	uint8_t offset[3];
+	/** NV_CONF register value. */
+	uint8_t nv_config;
+	/**
+	 * True if the sensor FIFO is currently enabled.
+	 *
+	 * Only headerless mode for the accelerometer alone is supported.
+	 */
+	bool fifo_enabled;
+	/** Pointer to data that will be read from the FIFO. */
+	const uint8_t *fifo_data;
+	/** Number of bytes remaining in fifo_data. */
+	uint16_t fifo_available;
+
+	/** True if in latched interrupt mode, otherwise non-latched. */
+	bool interrupt_mode_latched;
+	/** Raw value of INT1_IO_CTRL register. */
+	uint8_t int1_io_ctrl;
+	/** Raw value of INT_MAP_DATA register. */
+	uint8_t int_map_data;
+};
+
+struct bma4xx_emul_cfg {
+};
+
+void bma4xx_emul_reset(const struct emul *emul)
+{
+	struct bma4xx_emul_data *data = emul->data;
+
+	data->regs[BMA4XX_REG_CHIP_ID] = BMA4XX_CHIP_ID;
+	data->accel_enabled = false;
+	data->accel_range = 2;
+	data->odr_raw = 8;
+	memset(data->acceleration, 0, sizeof(data->acceleration));
+	memset(data->offset, 0, sizeof(data->offset));
+	data->nv_config = 0;
+	data->fifo_enabled = false;
+	data->fifo_data = NULL;
+	data->fifo_available = 0;
+	data->interrupt_mode_latched = false;
+}
+
+void bma4xx_emul_set_reg(const struct emul *target, uint8_t reg_addr, const uint8_t *val,
+			 size_t count)
+{
+	struct bma4xx_emul_data *data = target->data;
+
+	__ASSERT_NO_MSG(reg_addr + count <= NUM_REGS);
+	memcpy(data->regs + reg_addr, val, count);
+}
+
+void bma4xx_emul_get_reg(const struct emul *target, uint8_t reg_addr, uint8_t *val, size_t count)
+{
+	struct bma4xx_emul_data *data = target->data;
+
+	__ASSERT_NO_MSG(reg_addr + count <= NUM_REGS);
+	memcpy(val, data->regs + reg_addr, count);
+}
+
+bool bma4xx_emul_is_accel_enabled(const struct emul *emul)
+{
+	struct bma4xx_emul_data *data = emul->data;
+
+	return data->accel_enabled;
+}
+
+uint8_t bma4xx_emul_get_accel_range(const struct emul *emul)
+{
+	struct bma4xx_emul_data *data = emul->data;
+
+	return data->accel_range;
+}
```

Comment:
[remove] Checking the range can/should be done by either checking the register value or calling the driver's attribute getter.